<a href="https://colab.research.google.com/github/Ginkno/postech-tech-challenge-fase2-grupo1-12dtat/blob/Elizabeth/notebooks/03_modelagem.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 03 — Modelagem

**Dimensão 5 da rúbrica — 20 pontos.**

Mínimo de **dois** classificadores distintos. Um único modelo zera 8 dos 20 pontos.

In [1]:
from pathlib import Path

import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

# Semente fixa: use em TODO ponto com aleatoriedade (split, modelos, CV).
RANDOM_STATE = 42

RAW_1 = Path("/content") / "data" / "raw" / "application_record.csv"
RAW_2 = Path("/content") / "data" / "raw" / "credit_record.csv"

PROCESSED = Path("/content") / "data" / "processed" / "dataset_tratado.csv"
TARGET = "IS_BAD_PAYER"

pd.set_option("display.max_columns", None)

In [2]:
df_final_encoded = pd.read_csv(PROCESSED)

In [3]:
display(df_final_encoded.head())

,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,FLAG_MOBIL,FLAG_WORK_PHONE,FLAG_PHONE,FLAG_EMAIL,CNT_FAM_MEMBERS,YEARS_BIRTH,IS_INACTIVE,YEARS_EMPLOYED,TOTAL_MONTHS_OBSERVED,STATUS_LATED,STATUS_ON_TIME,STATUS_NO_CREDIT,MAX_RISK_SCORE,TOTAL_MONTHS_RECENT,STATUS_LATED_RECENT,STATUS_ON_TIME_RECENT,STATUS_NO_CREDIT_RECENT,MAX_RISK_SCORE_RECENT,TOTAL_MONTHS_OLDER,STATUS_LATED_OLDER,STATUS_NO_CREDIT_OLDER,MAX_RISK_SCORE_OLDER,IS_BAD_PAYER,INCOME_PER_PERSON,HAS_CHILDREN,NAME_INCOME_TYPE_Pensioner,NAME_INCOME_TYPE_State servant,NAME_INCOME_TYPE_Student,NAME_INCOME_TYPE_Working,NAME_EDUCATION_TYPE_Higher education,NAME_EDUCATION_TYPE_Incomplete higher,NAME_EDUCATION_TYPE_Lower secondary,NAME_EDUCATION_TYPE_Secondary / secondary special,NAME_FAMILY_STATUS_Married,NAME_FAMILY_STATUS_Separated,NAME_FAMILY_STATUS_Single / not married,NAME_FAMILY_STATUS_Widow,NAME_HOUSING_TYPE_House / apartment,NAME_HOUSING_TYPE_Municipal apartment,NAME_HOUSING_TYPE_Office apartment,NAME_HOUSING_TYPE_Rented apartment,NAME_HOUSING_TYPE_With parents,OCCUPATION_TYPE_Cleaning staff,OCCUPATION_TYPE_Cooking staff,OCCUPATION_TYPE_Core staff,OCCUPATION_TYPE_Drivers,OCCUPATION_TYPE_HR staff,OCCUPATION_TYPE_High skill tech staff,OCCUPATION_TYPE_IT staff,OCCUPATION_TYPE_Laborers,OCCUPATION_TYPE_Low-skill Laborers,OCCUPATION_TYPE_Managers,OCCUPATION_TYPE_Medicine staff,OCCUPATION_TYPE_Private service staff,OCCUPATION_TYPE_Realty agents,OCCUPATION_TYPE_Sales staff,OCCUPATION_TYPE_Secretaries,OCCUPATION_TYPE_Security staff,OCCUPATION_TYPE_Unknown,OCCUPATION_TYPE_Waiters/barmen staff
0,1,1,1,0,427500.0,1,1,0,0,2.0,32,0,12,16,2,13,1,2,12.0,0.0,12.0,0.0,0.0,4.0,2.0,1.0,2.0,0,213750.0,0,False,False,False,True,True,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False
1,1,1,1,0,427500.0,1,1,0,0,2.0,32,0,12,15,2,12,1,2,12.0,0.0,12.0,0.0,0.0,3.0,2.0,1.0,2.0,0,213750.0,0,False,False,False,True,True,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False
2,1,1,1,0,112500.0,1,0,0,0,2.0,58,0,3,30,7,7,16,1,12.0,3.0,7.0,2.0,1.0,18.0,4.0,14.0,1.0,0,56250.0,0,False,False,False,True,False,False,False,True,True,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False
3,0,0,1,0,270000.0,1,0,1,1,1.0,52,0,8,5,2,0,3,1,5.0,2.0,0.0,3.0,1.0,0.0,0.0,0.0,-2.0,0,270000.0,0,False,False,False,False,False,False,False,True,False,False,True,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False
4,0,0,1,0,270000.0,1,0,1,1,1.0,52,0,8,5,0,0,5,-1,0.0,0.0,0.0,0.0,-2.0,5.0,0.0,5.0,-1.0,0,270000.0,0,False,False,False,False,False,False,False,True,False,False,True,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False


## 1. Split treino/teste

`stratify` preserva a proporção das classes nos dois conjuntos.

In [4]:
df_model = df_final_encoded.copy()

from sklearn.model_selection import train_test_split

# Definir as features (X) e a variável alvo (y).
X = df_model.drop(columns=['IS_BAD_PAYER','MAX_RISK_SCORE_RECENT','MAX_RISK_SCORE'])
y = df_model['IS_BAD_PAYER']

#Dividindo os dados em 80% para treino e 20% para teste.
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=RANDOM_STATE
)

print(f'Shape de X_train: {X_train.shape}')
print(f'Shape de X_test: {X_test.shape}')
print(f'Shape de y_train: {y_train.shape}')
print(f'Shape de y_test: {y_test.shape}')

print('\nDistribuição da variável alvo no treino:')
display(y_train.value_counts(normalize=True))

print('\nDistribuição da variável alvo no teste:')
display(y_test.value_counts(normalize=True))

Shape de X_train: (26324, 62)
Shape de X_test: (6581, 62)
Shape de y_train: (26324,)
Shape de y_test: (6581,)

Distribuição da variável alvo no treino:


,proportion
IS_BAD_PAYER,
0,0.992592
1,0.007408



Distribuição da variável alvo no teste:


,proportion
IS_BAD_PAYER,
0,0.992554
1,0.007446


As variáveis 'MAX_RISK_SCORE_RECENT' e 'MAX_RISK_SCORE foram excluídas para evitar uma "cola" pelo modelo, já que a Target: 'IS_BAD_PAYER' foi definida baseada nos dados contidos nessas variáveis.

## 2. Modelos candidatos

Usar `Pipeline` evita vazamento: o scaler é ajustado só no fold de treino durante a validação cruzada.

Para este problema de classificação de detecção de maus pagadores, foram selecionados os seguintes modelos candidatos, que são algoritmos amplamente utilizados e robustos, especialmente em cenários com desbalanceamento de classes:

*   **Random Forest:** Um algoritmo ensemble baseado em árvores de decisão que se destaca pela boa performance, capacidade de lidar com dados complexos e redução de overfitting.
*   **Regressão Logística:** Um modelo linear simples, mas eficaz, frequentemente usado como linha de base. É interpretável e útil para entender a direção da relação entre as features e a variável alvo.
*   **XGBoost (Extreme Gradient Boosting):** Um algoritmo de boosting que é conhecido por sua alta performance e eficiência. É robusto a outliers e lida bem com dados desbalanceados através de parâmetros como `scale_pos_weight`.
*   **LightGBM (Light Gradient Boosting Machine):** Outro algoritmo de boosting, similar ao XGBoost, mas otimizado para velocidade e menor consumo de memória, mantendo uma performance competitiva. Também possui mecanismos para lidar com desbalanceamento de classes.

In [5]:
import numpy as np

# Calcular scale_pos_weight para lidar com o desbalanceamento de classes
neg_count = (y_train == 0).sum()
pos_count = (y_train == 1).sum()
scale_pos_weight_value = neg_count / pos_count

print(f"Scale Pos Weight calculado: {scale_pos_weight_value:.2f}")

Scale Pos Weight calculado: 133.99


In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, FunctionTransformer
from sklearn.compose import ColumnTransformer

import numpy as np

# --- Definindo as transformações de pré-processamento ---

# Transformação logarítmica para 'AMT_INCOME_TOTAL'
def log_transform(X):
    # Adiciona um pequeno valor para evitar log(0), se necessário,
    # mas np.log1p já trata isso
    return np.log1p(X)

log_transformer = FunctionTransformer(log_transform, validate=False)

# Capping para 'CNT_CHILDREN' e 'CNT_FAM_MEMBERS'
# Os valores de capping foram ajustados com base na análise
# de outliers no Notebook 01.
upper_bound_children = 2.50
upper_bound_family = 4.50

def cap_children(X):
    return np.clip(X, a_min=None, a_max=upper_bound_children)

def cap_family_members(X):
    return np.clip(X, a_min=None, a_max=upper_bound_family)

cap_children_transformer = FunctionTransformer(cap_children, validate=False)
cap_family_members_transformer = FunctionTransformer(cap_family_members, validate=False)

# Criando o ColumnTransformer para aplicar as transformações.
# 'remainder='passthrough'' garante que as outras colunas não sejam afetadas.
preprocessor = ColumnTransformer(
    transformers=[
        ('log_income', log_transformer, ['AMT_INCOME_TOTAL']),
        ('cap_children', cap_children_transformer, ['CNT_CHILDREN']),
        ('cap_family_members', cap_family_members_transformer, ['CNT_FAM_MEMBERS'])
    ],
    remainder='passthrough' # Mantenha as colunas não transformadas
)

# Definindo o passo de balanceamento (mantido para modelos que o usam)
# class_weight = 'balanced' (removido para modelos que não usam)

modelos = {
        "random_forest": Pipeline([
        ("preprocessor", preprocessor), # RF pode se beneficiar de features transformadas
        ("scaler", StandardScaler()), # É uma boa prática escalar mesmo para RF após transformações
        ("clf", RandomForestClassifier(n_estimators=300, class_weight='balanced',random_state=RANDOM_STATE, n_jobs=-1)),
    ]),
        "logistic_regression": Pipeline([
        ("preprocessor", preprocessor),
        ("scaler", StandardScaler()),
        ("clf", LogisticRegression(max_iter=1000, class_weight='balanced',random_state=RANDOM_STATE)),
    ]),
        "xgboost": Pipeline([
        ("preprocessor", preprocessor),
        ("scaler", StandardScaler()), # XGBoost pode se beneficiar de features escaladas, embora não seja estritamente necessário para árvores.
        ("clf", XGBClassifier(scale_pos_weight=scale_pos_weight_value,random_state=RANDOM_STATE, eval_metric='logloss')),
    ]),
    "lightgbm": Pipeline([
        ("preprocessor", preprocessor),
        ("scaler", StandardScaler()), # LightGBM também pode se beneficiar de features escaladas.
        ("clf", LGBMClassifier(scale_pos_weight=scale_pos_weight_value,random_state=RANDOM_STATE, n_jobs=-1)),
    ]),
}

## 3. Validação cruzada

In [7]:
from sklearn.model_selection import cross_val_score, StratifiedKFold

all_model_scores = {}

#Configura a validação cruzada estratificada (crucial para 0.74% de classe positiva)
cv_stratified = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)

print("Iniciando a Validação Cruzada Estratificada (Métrica: F1-Score)...\n")

for nome, modelo in modelos.items():
  # Passamos o cv_estratificado aqui para garantir que nenhum bloco fique sem a classe 1
    scores = cross_val_score(modelo, X_train, y_train, cv=cv_stratified, scoring="f1")
    all_model_scores[nome] = scores # Armazena os scores de cada fold
    print(f"{nome} - F1 Médio: {scores.mean().round(4)} (Desvio Padrão: {scores.std().round(4)})")

Iniciando a Validação Cruzada Estratificada (Métrica: F1-Score)...

random_forest - F1 Médio: 0.1223 (Desvio Padrão: 0.0614)
logistic_regression - F1 Médio: 0.0575 (Desvio Padrão: 0.0045)
xgboost - F1 Médio: 0.372 (Desvio Padrão: 0.0867)
[LightGBM] [Info] Number of positive: 156, number of negative: 20903
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.007427 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1093
[LightGBM] [Info] Number of data points in the train set: 21059, number of used features: 60
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.007408 -> initscore=-4.897792
[LightGBM] [Info] Start training from score -4.897792


/usr/local/lib/python3.13/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


[LightGBM] [Info] Number of positive: 156, number of negative: 20903
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.007630 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1074
[LightGBM] [Info] Number of data points in the train set: 21059, number of used features: 60
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.007408 -> initscore=-4.897792
[LightGBM] [Info] Start training from score -4.897792


/usr/local/lib/python3.13/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


[LightGBM] [Info] Number of positive: 156, number of negative: 20903
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.007717 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1068
[LightGBM] [Info] Number of data points in the train set: 21059, number of used features: 60
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.007408 -> initscore=-4.897792
[LightGBM] [Info] Start training from score -4.897792


/usr/local/lib/python3.13/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


[LightGBM] [Info] Number of positive: 156, number of negative: 20903
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.007367 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1088
[LightGBM] [Info] Number of data points in the train set: 21059, number of used features: 60
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.007408 -> initscore=-4.897792
[LightGBM] [Info] Start training from score -4.897792


/usr/local/lib/python3.13/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


[LightGBM] [Info] Number of positive: 156, number of negative: 20904
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.007577 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1072
[LightGBM] [Info] Number of data points in the train set: 21060, number of used features: 60
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.007407 -> initscore=-4.897840
[LightGBM] [Info] Start training from score -4.897840
lightgbm - F1 Médio: 0.2349 (Desvio Padrão: 0.1368)


/usr/local/lib/python3.13/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


## 4. Comparação

O modelo XGBoost apresentou a melhor performance entre todos os candidatos, com um F1-Score médio de 0.3720, superando significativamente o LightGBM (0.2349), o Random Forest (0.1223) e a Regressão Logística (0.0575). Dada a expressiva diferença de desempenho, o XGBoost foi selecionado como o modelo campeão para a etapa de otimização de hiperparâmetros, dispensando a necessidade de testes de hipótese adicionais.

 **Otimização de Hiperparâmetros**

Como o modelo XGBoost apresentou a melhor performance entre os outros modelos o próximo passo é otimizar seus hiperparâmetros. Isso é crucial para extrair o máximo de desempenho, garantindo que esteja configurado da melhor forma possível para o nosso conjunto de dados.

Optou-se pela utilização do RandomizedSearchCV devido ao elevado número de hiperparâmetros do modelo XGBoost. Essa abordagem permite explorar diferentes regiões do espaço de busca com menor custo computacional, mantendo boa capacidade de encontrar configurações de alto desempenho.

In [8]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import randint, uniform

# Parâmetros para RandomizedSearchCV para XGBoost
param_dist_xgb = {
    'clf__n_estimators': randint(100, 500), # Sorteia qualquer número inteiro entre 100 e 500
    'clf__learning_rate': uniform(0.01, 0.19), # Sorteia taxas decimais entre 0.01 e 0.20
    'clf__max_depth': randint(3, 10), # Sorteia profundidades inteiras entre 3 e 10
    'clf__subsample': uniform(0.6, 0.4), # entre 0.6 e 1.0
    'clf__colsample_bytree': uniform(0.6, 0.4), # entre 0.6 e 1.0
    'clf__gamma': uniform(0, 0.2)
}

cv_stratified = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)
xgb_pipeline = modelos['xgboost']

random_search_xgb = RandomizedSearchCV(
    estimator=xgb_pipeline,
    param_distributions=param_dist_xgb,
    n_iter=50, # Número de iterações
    cv=cv_stratified,
    scoring='f1',
    random_state=RANDOM_STATE,
    n_jobs=-1,
    verbose=1
)

print("Otimizando XGBoost...")
random_search_xgb.fit(X_train, y_train)

best_model = random_search_xgb.best_estimator_

print(f"O melhor modelo (best_estimator_):\n{best_model}")
print(f"Melhores hiperparâmetros para XGBoost: {random_search_xgb.best_params_}")
print(f"Melhor F1-score para XGBoost: {random_search_xgb.best_score_:.4f}")

Otimizando XGBoost...
Fitting 5 folds for each of 50 candidates, totalling 250 fits
O melhor modelo (best_estimator_):
Pipeline(steps=[('preprocessor',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('log_income',
                                                  FunctionTransformer(func=<function log_transform at 0x7f068d8176a0>),
                                                  ['AMT_INCOME_TOTAL']),
                                                 ('cap_children',
                                                  FunctionTransformer(func=<function cap_children at 0x7f068d56c400>),
                                                  ['CNT_CHILDREN']),
                                                 ('cap_family_members',
                                                  FunctionTransformer(func=<function cap_f...
                               grow_policy=None, importance_type=None,
                               interacti

O método avaliou diferentes combinações de parâmetros por meio de validação cruzada estratificada e selecionou automaticamente a configuração que apresentou melhor desempenho.

In [9]:
#Ganho obtido
ganho = ((0.4318 - 0.3720) / 0.3720) * 100
print(f"Ganho obtido: {ganho:.2f}%")

Ganho obtido: 16.08%


Como resultado o F1-Score aumentou de 0,3720 para 0,4318 representando um ganho aproximado de 16% na capacidade do modelo de identificar corretamente os maus pagadores.

## 5. Salvar Modelo e Dados de Teste para o Próximo Notebook

In [10]:
import joblib
from pathlib import Path

# Criar diretório para modelos se não existir
MODELS_DIR = Path("/content") / "models"
MODELS_DIR.mkdir(parents=True, exist_ok=True)

# Caminho para salvar o modelo
model_path = MODELS_DIR / "best_xgb_model.joblib"

# Salvar o best_model (o pipeline treinado)
joblib.dump(best_model, model_path)

print(f"Modelo salvo em: {model_path}")

Modelo salvo em: /content/models/best_xgb_model.joblib


In [11]:
from pathlib import Path

# Usar o diretório PROCESSED para salvar os dados de teste
PROCESSED_DIR = Path("/content") / "data" / "processed"
PROCESSED_DIR.mkdir(parents=True, exist_ok=True)

joblib.dump(X_test, PROCESSED_DIR / "X_test.joblib")
joblib.dump(y_test, PROCESSED_DIR / "y_test.joblib")

print("Dados de teste salvos com sucesso!")

Dados de teste salvos com sucesso!
